File: helm/relayforge/src/relayforge/api/backpressure.py
import asyncio


class Backpressure:
    """Мягкий release-scoped лимит активных delivery-Jobs + лимит одновременных
    create на один API-Pod.

    Проверка лимита и резервирование места под Job выполняются атомарно
    в JobRegistry.try_reserve (локальный per-Pod lock): в пределах одного Pod
    лимит соблюдается точно. Превышение возможно только между Pod'ами (HA)
    и ограничено числом Pod'ов — измеряется в сценарии 04. Подсчёт активных —
    локальный watch + pending, без cluster-wide list.
    """

    def __init__(self, max_active: int, max_concurrent_create: int, registry):
        self.max_active = max_active
        self.registry = registry
        self.create_semaphore = asyncio.Semaphore(max_concurrent_create)

    async def check_and_reserve(self, exempt_name: str | None) -> bool:
        """Проверяет лимит и резервирует место; True — место занято.

        Повтор ключа с уже известным Job (exempt_name известен registry)
        лимит не блокирует: вернёт False без резервирования.
        """
        if self.max_active <= 0:
            return False
        return await self.registry.try_reserve(self.max_active, exempt_name)

File: helm_golang/relayforge/internal/api/backpressure.go
package api

// Backpressure mirrors relayforge/api/backpressure.py: a per-pod semaphore
// (max concurrent creates) wrapping the registry's atomic check-and-reserve.
type Backpressure struct {
	maxActive           int
	registry            *JobRegistry
	createSemaphore     chan struct{}
}

// NewBackpressure mirrors the Python constructor: create_semaphore is an
// asyncio.Semaphore(max_concurrent_create) — modeled here as an unbuffered-ish
// counting gate of that size.
func NewBackpressure(maxActive, maxConcurrentCreate int, registry *JobRegistry) *Backpressure {
	return &Backpressure{
		maxActive:       maxActive,
		registry:        registry,
		createSemaphore: make(chan struct{}, maxConcurrentCreate),
	}
}

// acquire takes one slot of the concurrent-create semaphore.
func (b *Backpressure) acquire()    { b.createSemaphore <- struct{}{} }

// release returns one slot of the concurrent-create semaphore.
func (b *Backpressure) release() { <-b.createSemaphore }

// CheckAndReserve mirrors Backpressure.check_and_reserve: returns true when a
// slot was reserved; a known job (repeat key) returns false without reserving.
func (b *Backpressure) CheckAndReserve(exemptName string) (bool, error) {
	if b.maxActive <= 0 {
		return false, nil
	}
	return b.registry.TryReserve(b.maxActive, exemptName)
}